compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -287,14 +283,27 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                                     ),
                                 );
                             }
-                            if is_option_or_result && maybe_reinitialized_locations.is_empty() {
-                                err.span_suggestion_verbose(
-                                    fn_call_span.shrink_to_lo(),
-                                    "consider calling `.as_ref()` to borrow the type's contents",
-                                    "as_ref().".to_string(),
-                                    Applicability::MachineApplicable,
-                                );
+
+                            let borrow_method = match self_name {
+                                Some(sym::Option | sym::Result) => Some("as_ref"),
+                                Some(sym::Pin) => Some("as_mut"),
+                                _ => None,
+                            };
+
+                            if let Some(borrow_method) = borrow_method {
+                                if maybe_reinitialized_locations.is_empty() {
+                                    err.span_suggestion_verbose(
+                                        fn_call_span.shrink_to_lo(),
+                                        &format!(
+                                        "consider calling `.{}()` to borrow the type's contents",
```

Comment:
Extra nit, but I personally would love that level of precision in the diagnostics: in the case where `P = &mut _`, that is, when we have to deal with `Pin<&mut …>`, we already have a borrow; and what we need is a _**re**-borrow_.

  - This may look silly, but I think one of the main areas of confusion with `Pin` for beginners is this need for `.as_mut()`, precisely because it's a type which, contrary to `&mut`, features no implicit reborrowing.

    So even _an extra **note**_ for that very case, stating this very fact, could be quite educational:

    ```console
    note: this function takes ownership of the receiver `self`, which moves `foo`
      --> $DIR/pin-content-move.rs:6:12
       |
    LL |     fn foo(self: Pin<&mut Self>) {}
       |            ^^^^
    help: consider calling `.as_mut()` to reborrow the type's contents
       |
    LL |     foo.as_mut().foo();
       |         +++++++++
    note: this is needed because `Pin<&mut _>`, contrary to `&mut _`, does not feature implicit reborrowing.
       |
    LL |     foo.as_mut().foo();
       |         +++++++++
    ```

  - While that instance seems like a nit, in cases such as `Pin<Rc<…>>` _etc._ for shared-pinned stuff, the `.as_mut()` suggestion is also likely to be incorrect.

    So, if it weren't too annoying to implement, I'd do a heuristic whereby `Rc`, `Arc`, and `&mut _` would be special-cased, and would ask for, respectively: `.as_ref()` to _borrow_, `.as_ref()` to _borrow_, `.as_mut()` to _reborrow_, and otherwise defaulting to what we have here: `.as_mut()` to _borrow_.